mitmproxy/addons/proxyserver.py

```diff
@@ -121,15 +129,48 @@ async def shutdown_server(self):
         self.server = None
 
     async def handle_connection(self, r, w):
+        peername = w.get_extra_info('peername')
         asyncio_utils.set_task_debug_info(
             asyncio.current_task(),
             name=f"Proxyserver.handle_connection",
-            client=w.get_extra_info('peername'),
+            client=peername,
         )
         handler = ProxyConnectionHandler(
             self.master,
             r,
             w,
             self.options
         )
-        await handler.handle_client()
+        self._connections[peername] = handler
+        try:
+            await handler.handle_client()
+        finally:
+            del self._connections[peername]
+
+    def inject_event(self, flow: flow.Flow, event: events.Event):
+        if flow.client_conn.peername not in self._connections:
+            raise ValueError("Flow is not from a live connection.")
+        self._connections[flow.client_conn.peername].server_event(event)
+
+    @command.command("inject")
```

Comment:
How do you feel about having separate `inject_websocket` and `inject_tcp`? This makes it very explicit on the caller side and decreases the surface for errors.